File: Sharpex2D Java/framework/components/ComponentManager.java
package sharpex2d.framework.components;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import sharpex2d.framework.common.java.GenericHelper;

public class ComponentManager implements IComponent {
	
	/**
	 * Gets the Guid.
	 */
	public UUID getGuid()
	{
		return UUID.fromString("6A3D114D-6DF4-429E-82ED-F7CD0AE29CF8");
	}
	
	/**
	 * Constructs the component.
	 */
	public void construct()
	{
		for(int i = 0; i<= _components.size() - 1; i++)
		{
			IComponent component = _components.get(i);
			if(component instanceof IConstructable)
			{
				((IConstructable)component).construct();
			}
		}
		
		_isAlreadyConstructed = true;
	}

	/**
	 * Initializes a new ComponentManager class.
	 */
	public ComponentManager()
	{
		_components = new ArrayList<IComponent>();
	}
	
	private boolean _isAlreadyConstructed;
	
	private List<IComponent> _components;
	
	/**
	 * Adds a new Component.
	 * @param component The Component.
	 */
	public void addComponent(IComponent component)
	{
		_components.add(component);
		
		if(!_isAlreadyConstructed)
		{
			if(component instanceof IConstructable)
			{
				((IConstructable)component).construct();
			}
		}
	}
	/**
	 * Removes the Component.
	 * @param component The Component.
	 */
	public void removeComponent(IComponent component)
	{
		if(_components.contains(component))
		{
			_components.remove(component);
		}
	}
	
	@SuppressWarnings("unchecked")
	/**
	 * Gets the Component by guid.
	 * @param guid The Guid.
	 * @return Component.
	 */
	public <T> T getByGuid(UUID guid)
	{
		for(int i = 0; i<= _components.size() -1; i++)
		{
			IComponent component = _components.get(i);
			if(component.getGuid() == guid)
			{
				return (T)component;
			}
		}
		
		throw new IllegalArgumentException("The type if the guid " + guid + " could not be found.");
	}
	/**
	 * Gets the Component.
	 * @return Component.
	 */
	@SuppressWarnings("unchecked")
	public <T> T Get()
	{
		for(int i = 0; i<= _components.size() -1; i++)
		{
			if(GenericHelper.<T>getType() == _components.get(i).getClass())
			{
				return (T)_components.get(i);
			}
		}
		
		//try to query interfaces
		
		for(int i = 0; i<= _components.size() -1; i++)
		{
			if(QueryInterface(_components.get(i), GenericHelper.<T>getType()))
			{
				return (T)_components.get(i);
			}
		}
		
		throw new IllegalArgumentException("The type " + GenericHelper.<T>getType().toString() + " could not be found.");
	}
	/**
	 * Queries a Type.
	 * @param obj The Object.
	 * @param target The Target.
	 * @return Returns true if the object contains the specified class.
	 */
	@SuppressWarnings("rawtypes")
	private boolean QueryInterface(Object obj, Class target)
	{
		return target.isInstance(obj);
	}
}

File: Sharpex2D/Framework/Components/ComponentManager.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Sharpex2D.Framework.Components
{
    [Developer("ThuCommix", "developer@sharpex2d.de")]
    [Copyright("©Sharpex2D 2013 - 2014")]
    [TestState(TestState.Tested)]
    public class ComponentManager : IConstructable
    {
        #region IComponent Implementation

        /// <summary>
        ///     Sets or gets the Guid of the Component.
        /// </summary>
        public Guid Guid
        {
            get { return new Guid("6A3D114D-6DF4-429E-82ED-F7CD0AE29CF8"); }
        }

        #endregion

        private readonly List<IComponent> _internalComponents = new List<IComponent>();
        private bool _alreadyCalledConstruct;

        /// <summary>
        ///     Access to the Components enumeration.
        /// </summary>
        private List<IComponent> Components
        {
            get { return _internalComponents; }
        }

        /// <summary>
        ///     Initializes all Components.
        /// </summary>
        public void Construct()
        {
            for (int i = 0; i <= _internalComponents.Count - 1; i++)
            {
                var com = _internalComponents[i] as IConstructable;
                if (com != null)
                {
                    com.Construct();
                }
            }
            _alreadyCalledConstruct = true;
        }

        /// <summary>
        ///     Adds a new Component to the enumeration.
        /// </summary>
        /// <param name="component">The Component.</param>
        public void AddComponent(IComponent component)
        {
            Components.Add(component);
            if (_alreadyCalledConstruct)
            {
                //Single Construct.
                var com = component as IConstructable;
                if (com != null)
                {
                    com.Construct();
                }
            }
        }

        /// <summary>
        ///     Removes a Component from the enumeration.
        /// </summary>
        /// <param name="component">The Component.</param>
        public void RemoveComponent(IComponent component)
        {
            Components.Remove(component);
        }

        /// <summary>
        ///     Returns a specific component if exists.
        /// </summary>
        /// <typeparam name="T">The Type.</typeparam>
        /// <returns>Component</returns>
        public T Get<T>()
        {
            foreach (IComponent component in _internalComponents)
            {
                if (component == null) continue;
                if (component.GetType() == typeof (T) || component.GetType().BaseType == typeof(T))
                {
                    return (T) component;
                }
            }

            //if not found query interfaces 
            foreach (IComponent component in _internalComponents)
            {
                if (component == null) continue;
                if (QueryInterface(component.GetType(), typeof (T)))
                {
                    return (T) component;
                }
            }

            throw new InvalidOperationException("Component not found (" + typeof (T).FullName + ").");
        }

        /// <summary>
        ///     Gets the Component by Guid.
        /// </summary>
        /// <param name="guid">The Guid.</param>
        /// <returns>IComponent</returns>
        public IComponent GetByGuid(Guid guid)
        {
            foreach (IComponent component in _internalComponents)
            {
                if (component == null) continue;
                if (component.Guid == guid)
                {
                    return component;
                }
            }

            throw new InvalidOperationException("Component with guid " + guid + " not found.");
        }

        /// <summary>
        ///     Queries a type.
        /// </summary>
        /// <param name="type">The Type.</param>
        /// <param name="target">The TargetType.</param>
        /// <returns>True on success</returns>
        private bool QueryInterface(Type type, Type target)
        {
            return type.GetInterfaces().Any(implementation => implementation == target);
        }
    }
}